### ISSUE TITLE
Two-factor Authentication

### ISSUE THREAD

**Issue by insertish** [insertish/rauth#5]
Add different MFA methods.

### 1FA
- trusted hand over (USES MFA TICKET)

### 1FA (requires email specified)
- password
- email OTP
- security key

### 2FA (requires email + password specified)
- password + email OTP
- password + totp
- password + security key

**Comment by OrangeFlare** [comment]
Please make sure that you can't access your backup codes with your main account password ... Discord does this and a friend had an account stolen because of this flaw.

**Comment by stefanflegg** [comment]
> 
> 
> Please make sure that you can't access your backup codes with your main account password ... Discord does this and a friend had an account stolen because of this flaw.

That's very interesting, I didn't know that about Discord. I understood backup codes are typically generated and displayed once only at creation time and can't be retrieved afterwards.

**Comment by telmobss1** [comment]
please don't do like discord to authenticate through google authenticator, because for those who format a smartphone it's a drag

**Comment by bonn2** [comment]
> 
> 
> please don't do like discord to authenticate through google authenticator, because for those who format a smartphone it's a drag

Discord uses the standard TOTP format, you can use other apps that manage those codes *much* better than google authenticator. Authy is a popular one

**Comment by telmobss1** [comment]
but this creates a lot of problems for some people,


Email account, phone number, and sending the code via messages or to the verification email are sufficient

**Comment by bonn2** [comment]
@telmobss1 Codes sent like that are far less secure

**Comment by errhammr** [comment]
Does Security Key 2FA Login mean FIDO2/U2F/WebAuthn or the proprietary YubiKey OTP "standard"? I'd much rather prefer the former because it's an open standard that many security dongles support while the latter is only supported by dongles of the YubiKey brand.

**Comment by fredster33** [comment]
I think maybe what @telmobss1 is hinting at is to not *recommend* Google Authenticator like Discord. I agree, something OSS should be recommended, like Authy

**Comment by Chew** [comment]
@errhammr I'm hoping for Webauthn that seems the most like a protocol, and allows built-in peripherals like Apple's Touch ID or Face ID and even Windows Hello. Logging in with my phone 2FA is pain, but if all I need is a fingerprint, it would be very nice. More importantly, it needs to support more than one method (of course)

Notable platforms that do this: Cloudflare, GitHub, I'm sure the list goes on.

**Comment by henryistaken** [comment]
TOTP is a must and very happy to see it being considered! TOTP is meant to be an open standard, so it's probably best to avoid direct recommendations. "Scan with your authenticator app" is probably a better solution. 

> I think maybe what @telmobss1 is hinting at is to not recommend Google Authenticator like Discord. I agree, something OSS should be recommended, like Authy

Authy is not open source. Also please for the love of god let's not utilize an Authy-only TOTP solution like Twitch & Gemini have done. Just standard TOTP that is compatible with anything.

**Comment by fredster33** [comment]
> TOTP is a must and very happy to see it being considered! TOTP is meant to be an open standard, so it's probably best to avoid direct recommendations. "Scan with your authenticator app" is probably a better solution.
> 
> > I think maybe what @telmobss1 is hinting at is to not recommend Google Authenticator like Discord. I agree, something OSS should be recommended, like Authy
> 
> Authy is not open source. Also please for the love of god let's not utilize an Authy-only TOTP solution like Twitch & Gemini have done. Just standard TOTP that is compatible with anything.

My bad, Authy isn't open-source. This sounds even better, no direct reccomendations. We can probably link to a neutral list of auth apps (preferably only OSS ones) in case some users are just starting out with 2FA. I think that users should probably be nudged to 2FA as it's much more secure.

**Comment by person1234-ux** [comment]
> Please make sure that you can't access your backup codes with your main account password ... Discord does this and a friend had an account stolen because of this flaw.

I respectfully disagree please keep that feature. The fact that your friends discord account was stole is their fault not discord fault.

**Comment by Chew** [comment]
Discord is switching to requiring email confirmation for backup codes anyway...

Some services show you the backup codes once, but services like GitHub and Google let you view it whenever. Personally, I'm ambivalent, but it still needs to be easily accessible, if it is at all.

I do also disagree with the above user's attempt to use ad hominem to prove their point, it's still a valid point in terms of security. While it may have been caused by them, I still wouldn't put the blame on them, nor the fault of security on Discord's end (not that they particularly seem to care, despite them putting _some_ effort, it's still disappointingly little).

**Comment by LokeYourLord** [comment]
> > Please make sure that you can't access your backup codes with your main account password ... Discord does this and a friend had an account stolen because of this flaw.
> 
> I respectfully disagree please keep that feature. The fact that your friends discord account was stole is their fault not discord fault.

I disagree with your disagreement, had my account stolen the same way by an automated script due to an exploit that existed in Discord a while ago. Discord support being ass and completely unhelpful, having not done anything. So I agree with the first poster: Do NOT let anyone access backup codes without Email Confirmation or at all once displayed the first time.

**Comment by insertish** [comment]
Implemented MFA flow, now writing tests:
- [ ] https://github.com/insertish/rauth/issues/44
